hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/MorphingPortableReaderTest.java

```diff
@@ -0,0 +1,218 @@
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+/**
+ * MorphingPortableReader Tester.
+ *
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class MorphingPortableReaderTest {
+
+    private SerializationServiceV1 service1;
+    private SerializationServiceV1 service2;
+
+    private PortableReader reader;
+
+    @Before
+    public void before() throws Exception {
+        service1 = (SerializationServiceV1) new DefaultSerializationServiceBuilder()
+                .addPortableFactory(TestSerializationConstants.PORTABLE_FACTORY_ID, new PortableFactory() {
+                    public Portable create(int classId) {
+                        return new MorphingBasePortable();
+                    }
+
+                }).build();
+
+        service2 = (SerializationServiceV1) new DefaultSerializationServiceBuilder()
+                .addPortableFactory(TestSerializationConstants.PORTABLE_FACTORY_ID, new PortableFactory() {
+                    public Portable create(int classId) {
+                        return new MorphingPortable();
+                    }
+                }).build();
+
+        Data data = service1.toData(new MorphingBasePortable((byte) 1, true, (char) 2, (short) 3, 4, 5, 1f, 2d));
+
+        BufferObjectDataInput in = service2.createObjectDataInput(data);
+        PortableSerializer portableSerializer = service2.getPortableSerializer();
+        reader = portableSerializer.createMorphingReader(in);
+    }
+
+    @After
+    public void after() throws Exception {
+        service1.destroy();
+        service2.destroy();
+    }
+
+    @Test
+    public void testReadInt() throws Exception {
+        int aByte = reader.readInt("byte");
+        int aShort = reader.readInt("short");
+        int aChar = reader.readInt("char");
+        int aInt = reader.readInt("int");
+
+        assertEquals(1, aByte);
+        assertEquals(3, aShort);
+        assertEquals(2, aChar);
+        assertEquals(4, aInt);
+        assertEquals(0, reader.readInt("NO SUCH FIELD"));
+
+    }
+
+    @Test
+    public void testReadLong() throws Exception {
+        long aByte = reader.readLong("byte");
+        long aShort = reader.readLong("short");
+        long aChar = reader.readLong("char");
+        long aInt = reader.readLong("int");
+
+        assertEquals(1, aByte);
+        assertEquals(3, aShort);
+        assertEquals(2, aChar);
+        assertEquals(4, aInt);
+        assertEquals(0, reader.readLong("NO SUCH FIELD"));
+
+    }
+
+    @Test
+    public void testReadUTF() throws Exception {
+        assertNull(reader.readUTF("NO SUCH FIELD"));
+    }
+
+    @Test
+    public void testReadBoolean() throws Exception {
```

Comment:
Should the readBoolean not be tested on a boolean field? Because afaik it is only tested what happens when the field is not found, but we have no idea what happens when the boolean field is found.